File: packages/sdk/src/domains.ts
/** Poseidon2/BabyBear domains for protocol-v3 application statements. */
export const applicationDomains = Object.freeze({
  SCOPE: 0x10,
  NOTE: 0x11,
  NULLIFIER: 0x12,
  EMPTY_LEAF: 0x13,
  PAYOUT: 0x14,
  STATEMENT: 0x15,
  APP_MERKLE_NODE: 0x20,
} as const);

/** KeccakPair512 domains retained only by the proof system and parameter manifest. */
export const proofDomains = Object.freeze({
  PROOF_LEAF: 0x40,
  PROOF_NODE: 0x41,
  TRANSCRIPT_INIT: 0x42,
  TRANSCRIPT_ABSORB: 0x43,
  TRANSCRIPT_SQUEEZE: 0x44,
  PARAMETER_MANIFEST: 0x45,
} as const);

/** Keccak coordination labels for the v3 two-part proof protocol. */
export const coordinationDomains = Object.freeze({
  STATEMENT: "PQTC.V3.STATEMENT",
  PROOF: "PQTC.V3.PROOF",
  CHECKPOINT: "PQTC.V3.CHECKPOINT",
  VERIFICATION: "PQTC.V3.VERIFICATION",
} as const);

export const domains = Object.freeze({ ...applicationDomains, ...proofDomains });

export type ApplicationDomain = typeof applicationDomains[keyof typeof applicationDomains];
export type ProofDomain = typeof proofDomains[keyof typeof proofDomains];

export function isApplicationDomain(value: number): value is ApplicationDomain {
  return (value >= applicationDomains.SCOPE && value <= applicationDomains.STATEMENT)
    || value === applicationDomains.APP_MERKLE_NODE;
}

export function isProofDomain(value: number): value is ProofDomain {
  return value >= proofDomains.PROOF_LEAF && value <= proofDomains.PARAMETER_MANIFEST;
}

File: crates/pqtc-spec/src/domains.rs
//! Normative one-byte protocol domains. Changing any value changes the protocol.

pub const SCOPE: u8 = 0x10;
pub const NOTE: u8 = 0x11;
pub const NULLIFIER: u8 = 0x12;
pub const EMPTY_LEAF: u8 = 0x13;
pub const PAYOUT: u8 = 0x14;
pub const STATEMENT: u8 = 0x15;
pub const APP_MERKLE_NODE: u8 = 0x20;

pub const PROOF_LEAF: u8 = 0x40;
pub const PROOF_NODE: u8 = 0x41;
pub const TRANSCRIPT_INIT: u8 = 0x42;
pub const TRANSCRIPT_ABSORB: u8 = 0x43;
pub const TRANSCRIPT_SQUEEZE: u8 = 0x44;
pub const PARAMETER_MANIFEST: u8 = 0x45;
